### ISSUE TITLE
Cancellation question

### ISSUE THREAD

**Issue by halt-hammerzeit** [petkaantonov/bluebird#737]
Suppose we have such a Promise chain

```javascript
let i = 1
const action = () => asynchronous_console_log(i++)

const promise = action() // 1
.then(result => action()) // 2
.then
({
   action() // 3
   .then(result => action()) // this line is pending now (when the user calls .cancel())
   .then(result => action())
})
.then(result => action())

...

setTimeout
   promise.cancel()
```

Also suppose `action()` returns a cancellable Promise.
Suppose the current instruction is the second line of the inner promise chain, and the user calls `.cancel()`.
Will the inner promise chain get cancelled too?

I.e. will it count to `5` or will it count to `3`?

**Comment by bergus** [comment]
You've got a syntax error in line 7. Are you really passing an object literal to `then()`? If that was supposed to be an arrow function, did you intend to `return` the promise?

Yes, inner promise chains do get cancelled too.

No, in your example the whole chain will run before the timeout oocurs and you `.cancel()` the `promise`.

**Comment by benjamingr** [comment]
Please ask questions in the appropriate support channels.

**Comment by halt-hammerzeit** [comment]
@bergus It's not a syntax error. it's pseudocode.

So, are inner promise chains cancelled or not, when calling `.cancel()` on the parent thread?
It's the question to everyone and it remains open until anyone answers it (and proves it, e.g. a contributor or an author of the library).
Until someone answers we assume that `.cancel()` method doesn't know how to cancel inner promise chains?

**Comment by bergus** [comment]
@halt-hammerzeit Well, I have no idea then what your pseudocode does. Sorry. It would help a lot to provide syntactically valid (including ES6) code, because depending on what you mean the answer is different. Currently your code is ambigous.

Btw, I've already answered your question, but I'm not sure that our definitions of "*inner promise chain*" match.

**Comment by halt-hammerzeit** [comment]
@bergus We'll need more answers than just yours to be statistically more correct

**Comment by bergus** [comment]
@halt-hammerzeit  Go and read the code yourself if you want to be sure :-)

**Comment by halt-hammerzeit** [comment]
@bergus too difficult a task for me

**Comment by halt-hammerzeit** [comment]
@bergus Let the gurus or maintainers or the owner answer. They know for sure. Let's not mislead other people.

**Comment by bergus** [comment]
@halt-hammerzeit: Take my word as a guru or let it be :-) You won't got any other answers here as Benjamin pointed out, and if you use the [appropriate support channel](https://github.com/petkaantonov/bluebird#support) please make sure to post unambiguous code there.

**Comment by halt-hammerzeit** [comment]
@bergus I'd prefer the author of the library to answer himself.

**Comment by benjamingr** [comment]
@bergus is probably right, although it depends on whether or not `asynchronous_console_log` uses micro or macrotask semantics to resolve. If it's just a `Promise.resolve` all the code in the question will run before the `setTimeout` anyway.

Anyway, you can just test the code and see for yourself, it is generally best to avoid race conditions or edge cases in cancellation - it is being overhauled for 3.0 anyway by the way so the API will change when Petka is back with us :p

For what it's worth you can take typically take @bergus 's word. Please ask questions on IRC or stack overflow. This is not out of disrespect but SO helps their discoverability and keeps the issue tracker clean.

**Comment by halt-hammerzeit** [comment]
@benjamingr oh, actually, that's a good thought - I can make my sample code into a runnable test and then we'll know the 100% true answer.
I'll post back tomorrow.

**Comment by bergus** [comment]
Ah I see you've fixed your code example by inserting `result =>`. In this case, no, the inner promise chain is not cancelled because the outer chain doesn't know about it at all - your callback does not return a promise. You can fix this by inserting `return` or omitting the block braces.

**Comment by halt-hammerzeit** [comment]
@bergus Oh, sure, it's supposed to return there. I'm in the process of running the sample code now, I'll post back in 15 minutes.

**Comment by halt-hammerzeit** [comment]
So, I made the test and it counts to `3`.
Which means, as @bergus supposed, it cancels inner promise chains too.
That's a correct behaviour.
This can be closed now.

**Comment by halt-hammerzeit** [comment]
I updated the code in the first message and found out that it's not required to make the resulting promise `.cancellable()`: it's sufficient to call `.cancellable()`, for example, only on the first item of the promise chain.

**Comment by benjamingr** [comment]
@halt-hammerzeit I closed the question after commenting in the first message that this is not the correct support channel. I've reiterated that statement later again in another comment as did Bergi.

I'm locking this issue, you're creating noise for all subscribers to the tracker. You're welcome to continue discussion on IRC or via a Stack Overflow question which are the appropriate support channels.
